### PR TITLE
Including LinkingObjects in subscriptions

### PR DESCRIPTION
**Pull request by cmelchior**
Closes #https://github.com/realm/realm-java/issues/6426

API:
```
// Throws if called on non-query-based Realm
RealmQuery.includeLinkingObjects(String property)

// Usage: simple case
realm.where(Person.class)
  .includeLinkingObjects("parents")
  .subscribe("my-subscription");

// Usage: Advanced case with implicit subscription case
realm.where(Person.class)
  .equalTo("name", "Jane")
  .sort("age", Sort.ASCENDING)
  .includeLinkingObjects("parents", "other_parents")
  .findAllAsync();
```

I opted for the `RealmQuery.includeLinkingObjects()` method instead of e.g `RealmQuery.subscribe()` since this method already has 4 parameters in the worst case and we are having a combinatorial explosion trying to accommodate the many variants.  Also both `Clean Code` and `Effective Java` recommend either splitting up methods or introducing builders in that case. The method is stripped out when building without sync as well. 

TODO
- [x] Update changelog
- [x] Add new API
- [x] Tests

### CODE REVIEW

**Comment by ironage** [comment]
Couldn't the inclusions be part of the subscribe options along with the subscription name? That way there's no ambiguity about having inclusions on a local (non subscription) query. The js implementation does it this way.

**Comment by cmelchior** [comment]
Ready for review

File: realm/realm-library/src/main/cpp/io_realm_internal_core_IncludeDescriptor.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+#include "java_accessor.hpp"
+#include "java_query_descriptor.hpp"
+#include "util.hpp"
+
+
+using namespace realm;
+using namespace realm::util;
+using namespace realm::_impl;
+
+static void finalize_descriptor(jlong ptr);
```
Comment (kneth) [inline realm/realm-library/src/main/cpp/io_realm_internal_core_IncludeDescriptor.cpp]: Why do you have a prototype and an implementation?

File: realm/realm-library/src/main/cpp/io_realm_internal_core_IncludeDescriptor.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+{
+    TR_ENTER()
+    try {
+        JLongArrayAccessor table_arr(env, table_pointers);
+        JLongArrayAccessor index_arr(env, column_indexes);
+        auto starting_table = reinterpret_cast<Table*>(starting_table_ptr);
+        std::vector<LinkPathPart> parts;
+        parts.reserve(index_arr.size());
+        for (int i = 0; i < index_arr.size(); ++i) {
+            auto col_index = size_t(index_arr[i]);
```
Comment (kneth) [inline realm/realm-library/src/main/cpp/io_realm_internal_core_IncludeDescriptor.cpp]: `static_cast<size_t>(...)`

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 3]
```diff
@@ -2054,6 +2054,42 @@ private OsResults lazyFindAll() {
...
     }
 
+    /**
+     * This predicate is only relevant for Query-based Realms.
+     * <p>
+     * Objects referenced through fields marked with {@link io.realm.annotations.LinkingObjects} are normally not downloaded
+     * as part of the subscription in Query-based Realms, but by using this predicate, it is possible to specify which linking
+     * objects relationships should also be included in the subscription as well.
+     * <p>
+     * Note that all "forward" object references like object references and lists are always included in the subscription
```
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java]: Is it worth to mention that links and lists are followed eagerly?

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 4]
```diff
@@ -2054,6 +2054,42 @@ private OsResults lazyFindAll() {
         return this;
     }
 
+    /**
+     * This predicate is only relevant for Query-based Realms.
+     * <p>
+     * Objects referenced through fields marked with {@link io.realm.annotations.LinkingObjects} are normally not downloaded
```
Comment (kneth) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java]: If we have two classes `A` and `B`, and three objects `a1`, `a2`, and `b1`. If `a1` has a linking object to `b1`, and `b1` has a link to `a2`, all three objects will be included, when the query is on `A`, right?

Comment (cmelchior) [reply realm/realm-library/src/main/java/io/realm/RealmQuery.java]: Good question. I would assume yes.

Comment (kneth) [reply realm/realm-library/src/main/java/io/realm/RealmQuery.java]: I think it is important to document the behaviour. Maybe the API doc isn't the right place.

Comment (cmelchior) [reply realm/realm-library/src/main/java/io/realm/RealmQuery.java]: We should probably explain it in more details here: https://docs.realm.io/sync/using-synced-realms/syncing-data#subscribing-to-data-reading-data

**Pull request merged**